enterprise/server-enterprise/src/test/java/org/neo4j/server/rest/security/NeoFullRESTInteraction.java

```diff
@@ -0,0 +1,215 @@
+    }
+
+    @Override
+    public EnterpriseUserManager getManager()
+    {
+        return server.getDependencyResolver().resolveDependency( EnterpriseAuthManager.class ).getUserManager();
+    }
+
+    @Override
+    public String executeQuery( RESTSubject subject, String call, Map<String,Object> params,
+            Consumer<ResourceIterator<Map<String, Object>>> resultConsumer )
+    {
+        String escapedCall = call.replace( "'", "\\'" ).replace( "\"", "\\\"" );
+        HTTP.RawPayload payload = HTTP.RawPayload.quotedJson(
+                "{'statements':[{'statement':'" + escapedCall + "'}]}" );
+        HTTP.Response response = HTTP.withHeaders( HttpHeaders.AUTHORIZATION, subject.principalCredentials )
+                .request( POST, commitURL(), payload );
+
+        try
+        {
+            JsonNode data = JsonHelper.jsonNode( response.rawContent() );
+            if ( data.has( "errors" ) && data.get( "errors").has( 0 ) )
+            {
+                JsonNode firstError = data.get( "errors" ).get( 0 );
+                if ( firstError.has( "message" ) )
+                {
+                    return firstError.get( "message" ).asText();
+                }
+            }
+            if ( data.has( "results" ) && data.get( "results" ).has( 0 ) )
+            {
+                JsonNode firstResult = data.get( "results" ).get( 0 );
+                RESTResult result = new RESTResult( firstResult );
+                resultConsumer.accept( result );
+            }
+        }
+        catch ( JsonParseException e )
+        {
+            fail("Unexpected error parsing Json!");
+        }
+
+        return "";
+    }
+
+    @Override
+    public RESTSubject login( String username, String password ) throws Throwable
+    {
+        String principalCredentials = challengeResponse( username, password );
+        HTTP.Response response = authenticate( principalCredentials );
+        return new RESTSubject( response, principalCredentials );
+    }
+
+    private HTTP.Response authenticate( String principalCredentials )
+    {
+        return HTTP.withHeaders( HttpHeaders.AUTHORIZATION, principalCredentials ).request( POST, commitURL() );
+    }
+
+    @Override
+    public void logout( RESTSubject subject ) { }
+
+    @Override
+    public boolean isAuthenticated( RESTSubject subject )
+    {
+        subject.response = authenticate( subject.principalCredentials );
+        return subject.response.status() == 200 || // OK
+                subject.response.status() == 403; // PASSWORD CHANGE REQUIRED
+    }
+
+    @Override
+    public AuthenticationResult authenticationResult( RESTSubject subject )
+    {
+        switch ( subject.response.status() )
+        {
+            case 200: return AuthenticationResult.SUCCESS;
+            case 403: return AuthenticationResult.PASSWORD_CHANGE_REQUIRED;
+            case 429: return AuthenticationResult.TOO_MANY_ATTEMPTS;
+            default:  return AuthenticationResult.FAILURE;
```

Comment:
Sounds like a good idea. One option is to just enhance the NeoInteractionLevel interface with assertion methods like 'assertFailure' and 'assertSuccess', etc. And then how that is done is specific to embedded (uses AuthenticationResuilt), REST (uses status codes) and BOLT (uses other status codes). I would suggest we make this improvement as part of the work to support BOLT in these tests.